### PR TITLE
Implement new block reader for fuse engine

### PR DESCRIPTION
**Pull request by zhang2014**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Implement new block reader for fuse engine

## Changelog

- Improvement

## Related Issues

part #4345

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: query/src/storages/fuse/io/block_reader.rs [hunk 1]
```diff
@@ -15,121 +15,119 @@
...
+        let mut columns_array_iter = Vec::with_capacity(0);
+        for (index, column_meta) in &part.columns_meta {
+            let mut column_reader = self
+                .operator
+                .object(&part.location)
+                .range_reader(column_meta.offset, column_meta.length);
+
+            let mut column_chunk = vec![0; column_meta.length as usize];
+            // TODO: check thread name
+            column_reader.read_exact(&mut column_chunk).await?;
```
Comment (dantengsky) [inline query/src/storages/fuse/io/block_reader.rs:105]: does it mean that columns of a given block will be fetched sequentially?

File: query/src/storages/fuse/io/block_reader.rs [hunk 2]
```diff
@@ -15,121 +15,119 @@
...
+        let decompressor = BasicDecompressor::new(pages, vec![]);
+        Ok(column_iter_to_arrays(
+            vec![decompressor],
+            vec![descriptor_type],
+            field,
+            rows,
+        )?)
+    }
+
+    async fn read_columns(self, part: PartInfoPtr) -> Result<(usize, Vec<ArrayIter<'static>>)> {
```
Comment (dantengsky) [inline query/src/storages/fuse/io/block_reader.rs:91]: do we need some `tracing::instrument` here ?

**Comment by sundy-li** [comment]
```
MySQL [(none)]> SELECT Year, count(*) AS c FROM ontime_not_null GROUP BY Year ORDER BY c DESC;
ERROR 1105 (HY000): Code: 1002, displayText = missing field `col_metas` at line 1 column 10322.
MySQL [(none)]>
```

**Comment by dantengsky** [comment]
> ```
> MySQL [(none)]> SELECT Year, count(*) AS c FROM ontime_not_null GROUP BY Year ORDER BY c DESC;
> ERROR 1105 (HY000): Code: 1002, displayText = missing field `col_metas` at line 1 column 10322.
> MySQL [(none)]>
> ```

Oh,  the new filed `col_metas` is introduced in PR #4372, which is unfortunately not backward compatible.

hope after PR #4244 is done, data format migration will be backward compatible

**Comment by sundy-li** [comment]
I tested it and it's better than the previous version. (~50% improvement)

No more extra reads anymore.

**Comment by dantengsky** [comment]
UT / IT may have already found these issues:

on dataset `ontime`

~~~
mysql> SELECT DayOfWeek, count(*) AS c FROM ontime_not_null WHERE Year >= 2000 AND Year <= 2008 GROUP BY DayOfWeek ORDER BY c DESC;
ERROR 1105 (HY000): Code: 1058, displayText = downcast column error, column type: UInt8, expected column: "common_datavalues::columns::primitive::PrimitiveColumn<u16>".
~~~~

-----

~~~
2022-03-10T14:55:49.984720Z ERROR databend_query::servers::mysql::writers::query_result_writer: OnQuery Error: Code: 1058, displayText = downcast column error, column type: UInt8, expected column: "common_datavalu
es::columns::primitive::PrimitiveColumn<u16>".

   0: common_exception::exception_code::<impl common_exception::exception::ErrorCode>::UnknownColumn
             at common/exception/src/exception_code.rs:36:66
   1: common_datavalues::columns::series::Series::check_get::{{closure}}
             at common/datavalues/src/columns/series.rs:59:13
      core::option::Option<T>::ok_or_else
             at /rustc/8769f4ef2fe1efddd1f072485f97f568e7328f79/library/core/src/option.rs:1064:25
      common_datavalues::columns::series::Series::check_get
             at common/datavalues/src/columns/series.rs:58:19
   2: <common_functions::scalars::comparisons::comparison::ComparisonPrimitiveImpl<T,F> as common_functions::scalars::comparisons::comparison::ComparisonExpression>::eval
             at common/functions/src/scalars/comparisons/comparison.rs:278:48
   3: <common_functions::scalars::comparisons::comparison::ComparisonFunction as common_functions::scalars::function::Function>::eval
             at common/functions/src/scalars/comparisons/comparison.rs:96:19
   4: <common_functions::scalars::function_adapter::ArithmeticAdapter as common_functions::scalars::function::Function>::eval
             at common/functions/src/scalars/function_adapter.rs:347:9
   5: databend_query::pipelines::transforms::transform_expression_executor::ExpressionExecutor::execute_function
             at query/src/pipelines/transforms/transform_expression_executor.rs:185:22
      databend_query::pipelines::transforms::transform_expression_executor::ExpressionExecutor::execute
             at query/src/pipelines/transforms/transform_expression_executor.rs:110:45
   6: databend_query::pipelines::transforms::transform_filter::FilterTransform<_>::filter
             at query/src/pipelines/transforms/transform_filter.rs:67:28
      databend_query::pipelines::transforms::transform_filter::FilterTransform<_>::filter_map
             at query/src/pipelines/transforms/transform_filter.rs:72:15
   7: <databend_query::pipelines::transforms::transform_filter::FilterTransform<_> as databend_query::pipelines::processors::processor::Processor>::execute::{{closure}}::{{closure}}::{{closure}}
             at query/src/pipelines/transforms/transform_filter.rs:113:27
      <tokio_stream::stream_ext::filter_map::FilterMap<St,F> as futures_core::stream::Stream>::poll_next
~~~

**Comment by sundy-li** [comment]
cc @zhyass

File: query/src/storages/fuse/fuse_part.rs [hunk 3]
```diff
@@ -13,17 +13,36 @@
...
+use std::collections::HashMap;
 use std::sync::Arc;
 
 use common_exception::ErrorCode;
 use common_exception::Result;
 use common_planners::PartInfo;
 use common_planners::PartInfoPtr;
 
+#[derive(serde::Serialize, serde::Deserialize, PartialEq)]
+pub struct ColumnMeta {
```
Comment (sundy-li) [inline query/src/storages/fuse/fuse_part.rs:25]: Why not use the struct of `block::ColumnMeta`?

Comment (zhang2014) [reply query/src/storages/fuse/fuse_part.rs:25]: They may have different fields in the future.

**Review by sundy-li — APPROVED** [review]

File: query/src/storages/fuse/io/block_reader.rs (hunk 1 above)
Comment (zhang2014) [reply query/src/storages/fuse/io/block_reader.rs:110]: I will solve this in the next pr.

**Review by PsiACE — APPROVED** [review]

**Pull request merged**